andy/src/main/java/nl/tudelft/cse1110/andy/codechecker/checks/MockitoWhen.java

```diff
@@ -32,30 +33,59 @@ public MockitoWhen(String methodToSetUp, Comparison comparison, int expectedNumb
     public boolean visit(MethodInvocation mi) {
 
         String methodName = mi.getName().toString();
-
         /**
          * As soon as a Mockito's when happen, we wait for the next method call.
          * We know that a when() method was just called because of the inWhenMode variable flag.
          *
          * If the method call just after when() matches the one we are expecting, bingo!
          */
         if(inWhenMode) {
-            if(methodToSetUp.equals(methodName)) {
+            if(methodToSetUp.equals(methodName) || methodToSetUp.equals(previousMethodName)) {
+                // With the normal syntax, if we encounter a match in method name,
+                // it will be right after when(...).
+                // Example: when(mockedList.add("2")).thenReturn(true);
+                // the order in which they will be processed is thenReturn(...) -> when(...) -> add(...)
+                // With the other syntax we can encounter a match either before when or after it
+                // Example: doNothing().when(mockedList).someVoidMethod()
+                // the order will be someVoidMethod() -> when(...) -> doNothing()
+                // Alternatively, we can also write doReturn(true).when(mockedList.add("1"))
+                // the order will be when(...) -> doReturn(...) -> add(...)
+                // So we check if we encounter match on the current command or on the previous
                 numberOfCallsToWhen++;
+                inWhenMode = false;
+                previousMethodName = "";
+            } else if(previousMethodName.contains("do") && !methodName.equals("when")){
+                // If we did not encounter a match, and we are at a do* construct
+                // it means that we are going to the next statement and if it starts with "when",
+                // we do not reset inWhenMode. We clear previous to avoid potential bugs
+                inWhenMode = false;
+                previousMethodName = "";
+            }  else if(!methodName.contains("do") && !methodName.equals("when")){
```

Comment:
Do we this comparison with `when` and `do` in many places. Let's extract this to a private utility method like `isMockitoWhenMethod(...)`. Same for the `doNothing`.

Reply:
I agree that these conditions are sloppy and redundant, and I managed to reduce them to one method.